LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/DefaultClassCDevicesMessageSender.cs

```diff
@@ -12,84 +12,105 @@ namespace LoRaWan.NetworkServer
     public class DefaultClassCDevicesMessageSender : IClassCDeviceMessageSender
     {
         private readonly NetworkServerConfiguration configuration;
-        private readonly Region loRaRegion;
         private readonly ILoRaDeviceRegistry loRaDeviceRegistry;
         private readonly IPacketForwarder packetForwarder;
         private readonly ILoRaDeviceFrameCounterUpdateStrategyProvider frameCounterUpdateStrategyProvider;
 
         public DefaultClassCDevicesMessageSender(
             NetworkServerConfiguration configuration,
-            Region loRaRegion,
             ILoRaDeviceRegistry loRaDeviceRegistry,
             IPacketForwarder packetForwarder,
             ILoRaDeviceFrameCounterUpdateStrategyProvider frameCounterUpdateStrategyProvider)
         {
             this.configuration = configuration;
-            this.loRaRegion = loRaRegion;
             this.loRaDeviceRegistry = loRaDeviceRegistry;
             this.packetForwarder = packetForwarder;
             this.frameCounterUpdateStrategyProvider = frameCounterUpdateStrategyProvider;
         }
 
-        public async Task<bool> SendAsync(ILoRaCloudToDeviceMessage cloudToDeviceMessage, CancellationToken cts = default(CancellationToken))
+        public async Task<bool> SendAsync(IReceivedLoRaCloudToDeviceMessage cloudToDeviceMessage, CancellationToken cts = default(CancellationToken))
         {
             try
             {
                 if (string.IsNullOrEmpty(cloudToDeviceMessage.DevEUI))
                 {
-                    Logger.Log($"[C2D] DevEUI missing in payload", LogLevel.Error);
+                    Logger.Log($"[class-c] devEUI missing in payload", LogLevel.Error);
+                    return false;
+                }
+
+                if (!cloudToDeviceMessage.IsValid(out var validationErrorMessage))
+                {
+                    Logger.Log(cloudToDeviceMessage.DevEUI, $"[class-c] {validationErrorMessage}", LogLevel.Error);
                     return false;
                 }
 
                 var loRaDevice = await this.loRaDeviceRegistry.GetDeviceByDevEUIAsync(cloudToDeviceMessage.DevEUI);
                 if (loRaDevice == null)
                 {
-                    Logger.Log(cloudToDeviceMessage.DevEUI, $"[C2D] Device {cloudToDeviceMessage.DevEUI} not found", LogLevel.Error);
+                    Logger.Log(cloudToDeviceMessage.DevEUI, $"[class-c] device {cloudToDeviceMessage.DevEUI} not found", LogLevel.Error);
                     return false;
                 }
 
+                Region region;
+                switch (loRaDevice.Region)
```

Comment:
I think this would be nice to delegate to the RegionFactory. Create and pass in the loRaDevice.Region and you get the correct region back?